Default host user namespace via experimental flag

pweil- opened the pull request:
@vishh @ncdc @pmorie @smarterclayton @thockin 

Initial thought on the implementation https://github.com/kubernetes/kubernetes/pull/30684#issuecomment-241523425 wasn't quite right.  Since we need to dereference a PVC in some cases the defaulting code didn't fit nicely in the docker manager code (would've coupled it with a kube client and would've been messy).  I think passing this in via the runtime config turned out cleaner.  PTAL

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31169)

<!-- Reviewable:end -->

smarterclayton commented on cmd/kubelet/app/options/options.go:
```diff
@@ -211,4 +211,7 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
 	fs.BoolVar(&s.ProtectKernelDefaults, "protect-kernel-defaults", s.ProtectKernelDefaults, "Default kubelet behaviour for kernel tuning. If set, kubelet errors if any of kernel tunables is different than kubelet defaults.")
+
+	// experimental flags
+	fs.BoolVar(&s.ExperimentalHostUserNamespaceDefaulting, "experimental-host-user-namespace-defaulting", s.ExperimentalHostUserNamespaceDefaulting, "Default the value for UsernsMode for containers that are using other host namespaces, host mounts, or specific non-namespaced capabilities (MKNOD, SYS_MODULE, SYS_TIME). This should only be enabled is user namespace remapping is enabled in the docker d ...
```
In the comment, explain what you are defaulting the value to.  This description did not explain it.

smarterclayton commented on pkg/kubelet/container/runtime.go:
```diff
@@ -392,6 +392,9 @@ type RunContainerOptions struct {
 	ReadOnly bool
 	// hostname for pod containers
 	Hostname string
+	// EnableHostUserNamespace is true if it is recommended to use the host user namespace in the container
```
Comment should be: "When users request other host namespaces, automatically set userns=host for Docker containers" or similar.

smarterclayton commented on pkg/kubelet/kubelet.go:
```diff
@@ -899,6 +905,10 @@ type Kubelet struct {
 	// The bit of the fwmark space to mark packets for dropping.
 	iptablesDropBit int
+
+	// experimentalHostUserNamespaceDefaulting is an experimental flag to enable defaulting of the UsernsMode for
```
Better explanation here too.

pmorie commented on pkg/apis/componentconfig/types.go:
```diff
@@ -415,6 +415,9 @@ type KubeletConfiguration struct {
 	// Values must be within the range [0, 31]. Must be different from IPTablesMasqueradeBit
 	IPTablesDropBit int32 `json:"iptablesDropBit"`
+	// experimentalHostUserNamespaceDefaulting is an experimental flag to enable defaulting of the UsernsMode for
+	// a container.
```
Comment about this requiring remapping to be enabled in the daemon (or, ahem, container runtime?)

pmorie commented on pkg/kubelet/kubelet_hostusernamespace.go:
```diff
@@ -0,0 +1,90 @@
+// enableHostUserNamespace determines if the host user namespace should be used by the container runtime.
+// Returns true if the pod is using a host pid, pic, or network namespace, the pod is using a non-namespaced
+// capability, or the pod has a host path volume.
+func (kl *Kubelet) enableHostUserNamespace(pod *api.Pod) bool {
```
will you call this file `kubelet_pods.go`?  That's my next planned refactoring and I would just move this stuff into that file.

pmorie commented on pkg/kubelet/kubelet_hostusernamespace_test.go:
```diff
@@ -0,0 +1,212 @@
+/*
+Copyright 2015 The Kubernetes Authors.
```
2016

pmorie commented on pkg/kubelet/kubelet_hostusernamespace.go:
```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2015 The Kubernetes Authors.
```
2016

pmorie commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -0,0 +1,90 @@
+limitations under the License.
+*/
+
+package kubelet
```
note to other reviewers: I asked @pweil- to name this file as such, going to put a ton of other stuff into it in a planned refactor.

pmorie commented:
This LGTM, I'll give others a chance to chime in.

pmorie commented:
@smarterclayton @thockin is this going to be 1.4?

vishh commented:
Review status: 0 of 14 files reviewed at latest revision, 13 unresolved discussions.

---

*[cmd/kubelet/app/options/options.go, line 216 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31169#-KPsZBtQrM1qG0lpL3Mm:-KPsZBtQrM1qG0lpL3Mn:bkacd4e) ([raw file](https://github.com/kubernetes/kubernetes/blob/d6650fd437632eeef40a78dde590f8f479ab2f71/cmd/kubelet/app/options/options.go#L216)):*
> ```Go
> 
> 	// experimental flags
> 	fs.BoolVar(&s.ExperimentalHostUserNamespaceDefaulting, "experimental-host-user-namespace-defaulting", s.ExperimentalHostUserNamespaceDefaulting, "Default the value for UsernsMode for containers that are using other host namespaces, host mounts, or specific non-namespaced capabilities (MKNOD, SYS_MODULE, SYS_TIME). This should only be enabled is user namespace remapping is enabled in the docker daemon. [default=false]")
> ```

I'd prefer using `--feature-gates` flag to gate this feature instead of adding a special flag to it. 

---

*[pkg/apis/componentconfig/v1alpha1/types.go, line 475 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31169#-KPsZwQJHkchT3KankWM:-KPsZwQJHkchT3KankWN:bdivllh) ([raw file](https://github.com/kubernetes/kubernetes/blob/d6650fd437632eeef40a78dde590f8f479ab2f71/pkg/apis/componentconfig/v1alpha1/types.go#L475)):*
> ```Go
> 	// experimentalHostUserNamespaceDefaulting is an experimental flag to enable defaulting of the UsernsMode for
> 	// a container.
> 	ExperimentalHostUserNamespaceDefaulting bool `json:"experimentalHostUserNamespaceDefaulting"`
> ```

This feature should ideally be passed in via a `FeatureConfigs` field. @mtaufen is working on adding `FeatureConfigs` field to `kubeletConfiguration`.

---

*[pkg/kubelet/kubelet.go, line 253 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31169#-KPs__CPDlfFvLg81bxN:-KPs__CPDlfFvLg81bxO:bqxdzix) ([raw file](https://github.com/kubernetes/kubernetes/blob/d6650fd437632eeef40a78dde590f8f479ab2f71/pkg/kubelet/kubelet.go#L253)):*
> ```Go
> 	iptablesMasqueradeBit int,
> 	iptablesDropBit int,
> 	experimentalHostUserNamespaceDefaulting bool,
> ```

If you rebase, Kubelet factory now takes in a `kubeletConfiguration` object. You don't have to pass in this feature explicitly. 

---

*[pkg/kubelet/kubelet.go, line 1323 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31169#-KPsa8CmhYCAV0Fkcigk:-KPsa8CmhYCAV0Fkcigl:bt1tuoc) ([raw file](https://github.com/kubernetes/kubernetes/blob/d6650fd437632eeef40a78dde590f8f479ab2f71/pkg/kubelet/kubelet.go#L1323)):*
> ```Go
> 
> 	// only do this check if the experimental behavior is enabled, otherwise allow it to default to false
> 	if kl.experimentalHostUserNamespaceDefaulting {
> ```

Why not pass this feature down to the runtime and let the runtime shims take care of the defaulting?

---

*[pkg/kubelet/kubelet_pods.go, line 17 \[r6\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31169#-KPrRyiaPOHCnKYA147U-r6-17:-KPsaIFHlPcMwPXm4TN0:b-1z8qcu) ([raw file](https://github.com/kubernetes/kubernetes/blob/8dfdec5d9de07c74a414e048f43eddfa386650a1/pkg/kubelet/kubelet_pods.go#L17)):*
<details><summary><i>Previously, pmorie (Paul Morie) wrote…</i></summary>
> note to other reviewers: I asked @pweil- to name this file as such, going to put a ton of other stuff into it in a planned refactor.
</details>
Isn't this specific to docker runtime? Why keep this logic in the `kubelet` package?

---

*[pkg/kubelet/container/runtime.go, line 395 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31169#-KPoGsKER7jV7guPuBuT-r2-395:-KPsbiC-M84MyR8B-EbQ:bqxdjmj) ([raw file](https://github.com/kubernetes/kubernetes/blob/11725e4acd3837afd74603781cf7501015bb0ca1/pkg/kubelet/container/runtime.go#L395)):*
<details><summary><i>Previously, smarterclayton (Clayton Coleman) wrote…</i></summary>
> Comment should be: "When users request other host namespaces, automatically set userns=host for Docker containers" or similar.
</details>
I'd prefer making this change local to docker shim

---

*[pkg/kubelet/kubelet_hostusernamespace.go, line 38 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31169#-KPsbF3fI0ax0YaiF4-v:-KPsbF3fI0ax0YaiF4-w:bcppu26) ([raw file](https://github.com/kubernetes/kubernetes/blob/d6650fd437632eeef40a78dde590f8f479ab2f71/pkg/kubelet/kubelet_hostusernamespace.go#L38)):*
> ```Go
> func hasNonNamespacedCapability(pod *api.Pod) bool {
> 	for _, c := range pod.Spec.Containers {
> 		if c.SecurityContext != nil && c.SecurityContext.Capabilities != nil {
> ```

What about privileged pods?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31169)*
<!-- Sent from Reviewable.io -->

pweil- commented:
> Isn't this specific to docker runtime? Why keep this logic in the kubelet package?

it is specific to the docker runtime.  I just thought passing in a kube client to the runtime so that it could dereference the PVC was leaky so I didn't go that route. I could've put that part only in the kubelet and the rest of the code in docker_manager.go but then it was spread out more than it really needed to be (IMO).

pmorie commented:
> it is specific to the docker runtime.

These seem like perfectly reasonable criteria for me, regardless of container runtime.... what am i missing?

pweil- commented:
@vishh re: "If you rebase, Kubelet factory now takes in a kubeletConfiguration object. You don't have to pass in this feature explicitly."

do you mean you'd rather just set the value of this flag directly in `CreateAndInitKubelet` rather than passing it as an option to `NewMainKubelet`?  After rebase it looked like `CreateAndInitKubelet` was still passing all the args and did not include the whole config to that method so I'm not following

vishh commented:
Sorry, I was assuming that
https://github.com/kubernetes/kubernetes/pull/29216 has gone in. Looks like
its still in the queue :-( I'd prefer rebasing this PR over that because
that PR cleansup kubelet object initialization quite a bit.

On Tue, Aug 23, 2016 at 12:50 PM, Paul Weil <notifications@github.com>
wrote:

> @vishh <https://github.com/vishh> re: "If you rebase, Kubelet factory now
> takes in a kubeletConfiguration object. You don't have to pass in this
> feature explicitly."
>
> do you mean you'd rather just set the value of this flag directly in
> CreateAndInitKubelet rather than passing it as an option to NewMainKubelet?
> After rebase it looked like CreateAndInitKubelet was still passing all
> the args and did not include the whole config to that method so I'm not
> following
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31169#issuecomment-241854669>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKAUSfCXWVuxpeWJbUrcqDPzc310Sks5qi08VgaJpZM4JqWln>
> .
>

vishh commented:
The feature this PR is adding is specific to the docker runtime though. If
we are adding it all runtimes, then the CRI needs to change as well.

On Tue, Aug 23, 2016 at 12:01 PM, Paul Morie <notifications@github.com>
wrote:

> it is specific to the docker runtime.
>
> These seem like perfectly reasonable criteria for me, regardless of
> container runtime.... what am i missing?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31169#issuecomment-241840987>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKI8W8SnK8FonIsfl2jeMDeQIyyeBks5qi0OjgaJpZM4JqWln>
> .
>

vishh commented:
How about a `userns` package that can be consumed from dockertools package?

On Tue, Aug 23, 2016 at 11:55 AM, Paul Weil <notifications@github.com>
wrote:

> Isn't this specific to docker runtime? Why keep this logic in the kubelet
> package?
>
> it is specific to the docker runtime. I just thought passing in a kube
> client to the runtime so that it could dereference the PVC was leaky so I
> didn't go that route. I could've put that part only in the kubelet and the
> rest of the code in docker_manager.go but then it was spread out more than
> it really needed to be (IMO).
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31169#issuecomment-241838808>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKEOPd1PrhzF60ASIEMtPn8V0oEV1ks5qi0IKgaJpZM4JqWln>
> .
>

pweil- commented:
> How about a `userns` package that can be consumed from dockertools package?

I will need to update DockerManager to take a client directly (to give to the userns package) or indirectly (via a helper in the userns package).  It just seemed wrong that the container runtime would reach back to make api calls rather than just set a value based on config.  Since I hope this goes away shortly in favor of an api flag I don't mind, it just felt wrong as soon as I realized it needed to dereference the PVC to make a decision.

pmorie commented:
@vishh @pweil- I disagree that this PR is docker-specific entirely.  I think it's 95% runtime-agnostic plus the plumbing for docker.  In fact, rkt calls GenerateRunContainerOptions, so this seems exactly right to me.  What am I missing?  Am I an idiot (totally possible IMO)

pweil- commented:
> In fact, rkt calls GenerateRunContainerOptions

Yes, if wanted or needed to plumb this for another runtime this would be good.  I don't know if that's required or if it is, if it would be done before getting to an alpha api.  Obviously it's less work for me if we decide to leave it :+1:   Ultimately I think we key on an SC field like the other namespaces, have a flag to trigger this defaulting in the api server, and dump this code.

pweil- commented:
updated the commit to use feature gates.  I'll keep my eye on https://github.com/kubernetes/kubernetes/pull/29216 tonight and rebase when I can.

pmorie commented on pkg/util/config/feature_gate.go:
```diff
@@ -41,15 +41,22 @@ const (
+
+	// experimentalHostUserNamespaceDefaulting Default userns=host for containers
+	// that are using other host namespaces, host mounts, or specific non-namespaced capabilities
+	// (MKNOD, SYS_MODULE, SYS_TIME). This should only be enabled is user namespace remapping is enabled
```
nit: s/is/if/

pweil- commented:
@k8s-bot test this issue https://github.com/kubernetes/kubernetes/issues/30549

Commit:
experimental host user ns defaulting

pweil- commented:
rebased @vishh @pmorie @smarterclayton PTAL

manual test results and link to files for the defaulting scenarios just for completeness https://gist.github.com/pweil-/b13e719fb56dcdad74c2d421feae24b2

pmorie commented:
:eyes:

smarterclayton commented:
The changes related to checking the pod, applying the flag, and the tests look good to me.  I don't have any additional comments now that #29216 has gone in.

pmorie commented:
LGTM, will leave final sign-off to @vishh

ncdc commented on pkg/kubelet/container/runtime.go:
```diff
@@ -392,6 +392,11 @@ type RunContainerOptions struct {
 	ReadOnly bool
 	// hostname for pod containers
 	Hostname string
+	// EnableHostUserNamespace sets userns=true when users request host namespaces (pid, ipc, net),
```
Do you mean userns=host?

pweil- replied:
yup.  fixed.

vishh commented:
Review status: 0 of 6 files reviewed at latest revision, 15 unresolved discussions.

---

*[pkg/kubelet/kubelet_pods.go, line 38 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31169#-KPsbF3fI0ax0YaiF4-v:-KQ7KrAwqq2DHAW5223v:bpo0s0l) ([raw file](https://github.com/kubernetes/kubernetes/blob/d6650fd437632eeef40a78dde590f8f479ab2f71/pkg/kubelet/kubelet_pods.go#L38)):*
<details><summary><i>Previously, vishh (Vish Kannan) wrote…</i></summary>
> What about privileged pods?
</details>
@pweil- Is it assumed that privileged pods are always run with initial userns?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31169)*
<!-- Sent from Reviewable.io -->

vishh commented:
Review status: 0 of 6 files reviewed at latest revision, 15 unresolved discussions.

---

*[cmd/kubelet/app/options/options.go, line 216 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31169#-KPoGsKCh2eDEsv5n5Vg-r2-216:-KQ7LYT8hjRB1lExr9am:b16p1xv) ([raw file](https://github.com/kubernetes/kubernetes/blob/11725e4acd3837afd74603781cf7501015bb0ca1/cmd/kubelet/app/options/options.go#L216)):*
<details><summary><i>Previously, smarterclayton (Clayton Coleman) wrote…</i></summary>
> In the comment, explain what you are defaulting the value to.  This description did not explain it.
</details>
It is not specific to docker daemon. So why mention docker in the description?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31169)*
<!-- Sent from Reviewable.io -->

pwittrock commented:
Just to follow up, this is NOT intended for the v1.4 milestone, correct?

pwittrock commented:
cc @pwittrock

pweil- commented:
> @pweil- Is it assumed that privileged pods are always run with initial userns?

Based on https://github.com/mrunalp/docker/blob/3c3d2bf85229f9ca4879a24072762477e96c4344/daemon/daemon_unix.go#L486-L489 a privileged container that is using private user namespaces when remapping is enabled will be rejected by the daemon.  Added it to the defaulting  mechanism.

@mrunalp do you agree with this?

pweil- commented:
> @mrunalp do you agree with this?

Spoke with @mrunalp.  There may be a deeper issue with privileged=true when remap is on and userns is != "host".  He is investigating to see why it is rejected in the code I referenced.  

We should wait on that before saying go/no-go on defaulting privileged containers.

pweil- commented:
> Just to follow up, this is NOT intended for the v1.4 milestone, correct?

I was hoping to squeak it in but based on the last comment I'd say we're officially stalled and I know we're past cutoff anyway :sob:

mrunalp commented:
docker doesn't allow privileged with userns as privileged implies less security and userns adds more security. I'll open an issue and follow up with docker to see if docker can enable privileged with user namespaces. So, there are two options for kubernetes:
1. privileged=host implies userns=host
2. privileged=host could be supported with userns=private by tweaking the container settings in kubernetes that have the same effect as privileged in docker and if at a later date docker allows it then rely on docker --privileged to do it.

pweil- commented:
> 1. privileged=host implies userns=host

covered by the new defaulting logic I added - ready for review

> 2. privileged=host could be supported with userns=private by tweaking the container settings in kubernetes that have the same effect as privileged in docker and if at a later date docker allows it then rely on docker --privileged to do it.

probably something we should discuss in the feature proposal if this is the route we want to go

Thanks for checking up on this @mrunalp!

pmorie commented:
@vishh - how do you feel about this going into 1.4?

vishh commented:
If there is no urgency for this feature, we can avoid merging it into v1.4.
We can always say that the next v1.5 alpha release will have this feature.

On Fri, Aug 26, 2016 at 3:03 PM, Paul Morie <notifications@github.com>
wrote:

> @vishh <https://github.com/vishh> - how do you feel about this going into
> 1.4?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31169#issuecomment-242862670>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKLe2faaXnhzL_4-7WxUSqy8W9m5hks5qj2KagaJpZM4JqWln>
> .
>

smarterclayton commented:
There is some urgency to be able to enable this sooner rather than
later, so we can begin testing and stabilizing user namespaces for use
with Kube.

Increasing the security of the container runtime is of high importance
and urgency to us.  This change is fairly minimal and gated behind a
flag.

derekwaynecarr commented:
I am +1 on this change, just ask we have a TODO in dockershim that says add support for this so its not lost in the transition to the CRI.

pweil- commented:
> I am +1 on this change, just ask we have a TODO in dockershim that says add support for this so its not lost in the transition to the CRI.

Added TODO in dockershim service file about environment detection that could enable defaulting in addition to any introduction of new API features.

dchen1107 commented:
cc/ @pwittrock, our release czar

ncdc commented:
@k8s-bot node e2e test this issue: #27462

pweil- commented:
> 1) What is the worst case scenario in accepting the change - what are the risks?

Risks associated with this feature:  

1.  that the behavior simply does not work.  Since this feature is gated by a flag this is mitigated by turning the feature off and we revert to the same state we are in now
2.  the behavior defaults incorrectly.  If incorrect defaulting occurs a container may run with the host namespace unexpectedly.  This is also mitigated by the flag, the pod would need to be deleted and recreated in this case.
3.  the behavior is all or nothing - when enabled, defaulting will occur for any container that meets the conditions outlined in this PR.  I consider this a risk because a "normal" user can launch a container with the host user namespace.  However, this is the case for other privileged operations in the absence of PSP controls so this is an acceptable risk.

> 2) How difficult is it to undo the change if we discover an issue before the release - a simple rollback.

Rollback the PR, this is self contained.

> 3) What critical value does this give that is needed for v1.4?

There is currently no way to configure a container to run with the host user namespace in Kubernetes.  This means that if an admin would like to further secure their cluster by enabling remapping they will not be able to launch more privileged containers that need access to the host namespace.

derekwaynecarr commented:
@pwittrock - I have moved this to the 1.5 milestone per our discussion.

@dchen1107 @vishh - can you LGTM this so it can merge when 1.5 is opened given the consensus we reached in sig-node?  I am fine deferring this to the next release, but would like acknowledgment in case we carry an interim patch downstream.  Thanks!

vishh commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -0,0 +1,103 @@
+}
+
+// hasPrivilegedContainer returns true if any of the containers in the pod are privileged.
+func hasPrivilegedContainer(pod *api.Pod) bool {
```
Why does the entire pod have to run in the native userns if a single container is privileged?

pweil- replied:
I worry about the interaction between mixed namespaces in a single pod.  We've seen issues in the past with them that were very hard to debug.  Also, the flags for controlling namespaces are on the host level and get applied to every container so I went with that convention here.

pweil- replied:
> Also, the flags for controlling namespaces are on the host level and get applied to every container so I went with that convention here.

I should say, specifically, the flags are on the `PodSecurityContext`.

smarterclayton replied:
Yeah I thought this was consistent with the other flags.  Specifically net
(you can't escape net in just one container).

Although as soon as I typed that, I realized that we have a use case for
privileged in one container and not in another coming up which now has me
terribly worried.

On Fri, Sep 2, 2016 at 2:52 PM, Paul Weil <notifications@github.com> wrote:

> In pkg/kubelet/kubelet_pods.go
> <https://github.com/kubernetes/kubernetes/pull/31169#discussion_r77395627>
> :
>
> > +	"k8s.io/kubernetes/pkg/api"
> > +)
> > +
> > +// enableHostUserNamespace determines if the host user namespace should be used by the container runtime.
> > +// Returns true if the pod is using a host pid, pic, or network namespace, the pod is using a non-namespaced
> > +// capability, the pod contains a privileged container, or the pod has a host path volume.
> > +func (kl *Kubelet) enableHostUserNamespace(pod *api.Pod) bool {
> > +	if hasPrivilegedContainer(pod) || hasHostNamespace(pod) ||
> > +		hasHostVolume(pod) || hasNonNamespacedCapability(pod) || kl.hasHostMountPVC(pod) {
> > +		return true
> > +	}
> > +	return false
> > +}
> > +
> > +// hasPrivilegedContainer returns true if any of the containers in the pod are privileged.
> > +func hasPrivilegedContainer(pod *api.Pod) bool {
>
> Also, the flags for controlling namespaces are on the host level and get
> applied to every container so I went with that convention here.
>
> I should say, specifically, the flags are on the PodSecurityContext.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31169/files/423ccec9b7b7582c3a4298015d4d23a71d664567#r77395627>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p5xAH6Vm_YI1XPXdmCfGMBQLe9YJks5qmHCBgaJpZM4JqWln>
> .
>

vishh replied:
AFAIK, we don't share userns in a pod. So why would it matter if a single
container in a multi-container pod were to use native userns?

On Fri, Sep 2, 2016 at 1:08 PM, Clayton Coleman <notifications@github.com>
wrote:

> In pkg/kubelet/kubelet_pods.go
> <https://github.com/kubernetes/kubernetes/pull/31169#discussion_r77404766>
> :
>
> > +	"k8s.io/kubernetes/pkg/api"
> > +)
> > +
> > +// enableHostUserNamespace determines if the host user namespace should be used by the container runtime.
> > +// Returns true if the pod is using a host pid, pic, or network namespace, the pod is using a non-namespaced
> > +// capability, the pod contains a privileged container, or the pod has a host path volume.
> > +func (kl *Kubelet) enableHostUserNamespace(pod *api.Pod) bool {
> > +	if hasPrivilegedContainer(pod) || hasHostNamespace(pod) ||
> > +		hasHostVolume(pod) || hasNonNamespacedCapability(pod) || kl.hasHostMountPVC(pod) {
> > +		return true
> > +	}
> > +	return false
> > +}
> > +
> > +// hasPrivilegedContainer returns true if any of the containers in the pod are privileged.
> > +func hasPrivilegedContainer(pod *api.Pod) bool {
>
> Yeah I thought this was consistent with the other flags. Specifically net
> (you can't escape net in just one container). Although as soon as I typed
> that, I realized that we have a use case for privileged in one container
> and not in another coming up which now has me terribly worried.
> … <#m_-3924339393635099543_>
> On Fri, Sep 2, 2016 at 2:52 PM, Paul Weil ***@***.***> wrote: In
> pkg/kubelet/kubelet_pods.go <#31169 (comment)
> <https://github.com/kubernetes/kubernetes/pull/31169#discussion_r77395627>>
> : > + "k8s.io/kubernetes/pkg/api" > +) > + > +// enableHostUserNamespace
> determines if the host user namespace should be used by the container
> runtime. > +// Returns true if the pod is using a host pid, pic, or network
> namespace, the pod is using a non-namespaced > +// capability, the pod
> contains a privileged container, or the pod has a host path volume. > +func
> (kl *Kubelet) enableHostUserNamespace(pod *api.Pod) bool { > + if
> hasPrivilegedContainer(pod) || hasHostNamespace(pod) || > +
> hasHostVolume(pod) || hasNonNamespacedCapability(pod) ||
> kl.hasHostMountPVC(pod) { > + return true > + } > + return false > +} > + >
> +// hasPrivilegedContainer returns true if any of the containers in the pod
> are privileged. > +func hasPrivilegedContainer(pod *api.Pod) bool { Also,
> the flags for controlling namespaces are on the host level and get applied
> to every container so I went with that convention here. I should say,
> specifically, the flags are on the PodSecurityContext. — You are receiving
> this because you were mentioned. Reply to this email directly, view it on
> GitHub <https://github.com/kubernetes/kubernetes/pull/31169/files/
> 423ccec9b7b7582c3a4298015d4d23a71d664567#r77395627>, or mute the thread <
> https://github.com/notifications/unsubscribe-auth/ABG_p5xAH6Vm_
> YI1XPXdmCfGMBQLe9YJks5qmHCBgaJpZM4JqWln> .
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31169/files/423ccec9b7b7582c3a4298015d4d23a71d664567#r77404766>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKMyft_lxbDu_hDB2dsGOTKwmfxqEks5qmIIsgaJpZM4JqWln>
> .
>

mrunalp replied:
> AFAIK, we don't share userns in a pod. So why would it matter if a single container in a multi-container pod were to use native userns?

Actually, it is required that a container shares a userns with another container if it shares any namespace with the other container else it won't have the correct capabilities in the namespace. (This happens by default in docker).

vishh replied:
Ohhh!!! Does this mean that with userns enabled, containers in a pod will
automatically share userns with docker?

On Fri, Sep 2, 2016 at 2:19 PM, Mrunal Patel <notifications@github.com>
wrote:

> In pkg/kubelet/kubelet_pods.go
> <https://github.com/kubernetes/kubernetes/pull/31169#discussion_r77413259>
> :
>
> > +	"k8s.io/kubernetes/pkg/api"
> > +)
> > +
> > +// enableHostUserNamespace determines if the host user namespace should be used by the container runtime.
> > +// Returns true if the pod is using a host pid, pic, or network namespace, the pod is using a non-namespaced
> > +// capability, the pod contains a privileged container, or the pod has a host path volume.
> > +func (kl *Kubelet) enableHostUserNamespace(pod *api.Pod) bool {
> > +	if hasPrivilegedContainer(pod) || hasHostNamespace(pod) ||
> > +		hasHostVolume(pod) || hasNonNamespacedCapability(pod) || kl.hasHostMountPVC(pod) {
> > +		return true
> > +	}
> > +	return false
> > +}
> > +
> > +// hasPrivilegedContainer returns true if any of the containers in the pod are privileged.
> > +func hasPrivilegedContainer(pod *api.Pod) bool {
>
> AFAIK, we don't share userns in a pod. So why would it matter if a single
> container in a multi-container pod were to use native userns?
>
> Actually, it is required that a container shares a userns with another
> container if it shares any namespace with the other container else it won't
> have the correct capabilities in the namespace. (This happens by default in
> docker).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31169/files/423ccec9b7b7582c3a4298015d4d23a71d664567#r77413259>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKGUbkJJZLbPYiXhYhlFA9xq0AxoFks5qmJLMgaJpZM4JqWln>
> .
>

mrunalp replied:
Yes.

> On Sep 2, 2016, at 3:43 PM, Vish Kannan <notifications@github.com> wrote:
> 
> In pkg/kubelet/kubelet_pods.go:
> 
> > +	"k8s.io/kubernetes/pkg/api"
> > +)
> > +
> > +// enableHostUserNamespace determines if the host user namespace should be used by the container runtime.
> > +// Returns true if the pod is using a host pid, pic, or network namespace, the pod is using a non-namespaced
> > +// capability, the pod contains a privileged container, or the pod has a host path volume.
> > +func (kl *Kubelet) enableHostUserNamespace(pod *api.Pod) bool {
> > +	if hasPrivilegedContainer(pod) || hasHostNamespace(pod) ||
> > +		hasHostVolume(pod) || hasNonNamespacedCapability(pod) || kl.hasHostMountPVC(pod) {
> > +		return true
> > +	}
> > +	return false
> > +}
> > +
> > +// hasPrivilegedContainer returns true if any of the containers in the pod are privileged.
> > +func hasPrivilegedContainer(pod *api.Pod) bool {
> Ohhh!!! Does this mean that with userns enabled, containers in a pod will automatically share userns with docker?
> …
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

pweil- replied:
I think this is resolved then.  Thanks @mrunalp for clarifying

vishh replied:
Ok. Then all containers in k8s should be sharing userns since they share netns by default. 
@pweil- Can you include this information as a comment in your PR?

pweil- replied:
added
```
// NOTE: when if a container shares any namespace with another container it must also share the user namespace
// or it will not have the correct capabilities in the namespace.  This means that host user namespace
// is enabled per pod, not per container.
```
to `kubelet_pods.go`

vishh commented:
@derekwaynecarr Ack! One more comment needs to be addressed and then this PR is good to go

vishh commented:
LGTM excepting one nit - https://github.com/kubernetes/kubernetes/pull/31169#discussion_r77688628

feiskyer commented:
@yifan-gu Does rkt support this feature?

yifan-gu commented:
@feiskyer By default, it's running in host user namespace, unless `--private-users` flag is set.

vishh commented:
@pweil- Were you intending to close this PR?

pweil- commented:
> @pweil- Were you intending to close this PR?

I thought the consensus was that we were going to get this in for experimental support for escaping user ns while work was continued on a more well rounded solution.  I can close if that is a misunderstanding on my part.

vishh commented:
I was under the impression that this experimental feature is being dropped
in favor of a more general solution for v1.5. If that is not the case, let
get this PR merged.

On Tue, Oct 11, 2016 at 6:23 AM, Paul Weil <notifications@github.com> wrote:

> @pweil- <https://github.com/pweil-> Were you intending to close this PR?
>
> I thought the consensus was that we were going to get this in for
> experimental support for escaping user ns while work was continued on a
> more well rounded solution. I can close if that is a misunderstanding on my
> part.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31169#issuecomment-252914695>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKJN33baY-Z1L0rvk-_Pll_55sKEtks5qy43BgaJpZM4JqWln>
> .
>

vishh commented on pkg/kubelet/kubelet.go:
```diff
@@ -493,6 +493,14 @@ func NewMainKubelet(kubeCfg *componentconfig.KubeletConfiguration, kubeDeps *Kub
 		iptablesMasqueradeBit:        int(kubeCfg.IPTablesMasqueradeBit),
 		iptablesDropBit:              int(kubeCfg.IPTablesDropBit),
+
+		// TODO: this can be removed and changed to use the feature gate when it is added to the kubelet
```
nit: Feature Gates is already available in the kubelet

vishh reviewed (changes requested):
@pweil- Let's use feature gates flag to control this feature. Otherwise this PR is good to go.

pweil- commented:
@vishh this was actually already using featuregate to enable the behavior in the Kubelet so I removed that comment.  Please let me know if that is incorrect.

Rebased

pweil- commented:
@k8s-bot node e2e test this

vishh commented:
@pweil- I'd like clarify the support for `experimental` features. In my mind, `experimental` features at the node level can either be,

1. Graduated to long term support, or
2. Superseded by a well though out or generic feature/implementation.
3. Deleted (if necessary).

As for this PR, I'm hoping that we will implement a more generic solution soon. WDYT?

derekwaynecarr commented:
To be fair, I think (3) should never apply to this use case.  I see no
reason to remove the experimental support without a viable alternative.

On Wednesday, November 2, 2016, Vish Kannan <notifications@github.com>
wrote:

> @pweil- <https://github.com/pweil-> I'd like clarify the support for
> experimental features. In my mind, experimental features at the node
> level can either be,
>
>    1. Graduated to long term support, or
>    2. Superseded by a well though out or generic feature/implementation.
>    3. Deleted (if necessary).
>
> As for this PR, I'm hoping that we will implement a more generic solution
> soon. WDYT?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31169#issuecomment-258002306>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF8dbA_pKOUbdb5PPMQGMFnoQYn-sckEks5q6P5jgaJpZM4JqWln>
> .
>

pweil- commented:
@vishh I think there will be a combination of both graduation, possibly api level support, and likely a more robust way for the kubelet to manage permissions.  The options I see are drawn out in https://github.com/kubernetes/kubernetes/pull/34569 but haven't received much feedback but if we can get consensus on what the path forward is I think we can probably get the work done in 1.6.

pweil- commented:
rebased

smarterclayton commented:
Reapplying labels - anyone who wants to use host namespaces either needs this, or needs the follow up to it (and probably both, because the behavior change is not necessarily backwards compatible in a generic solution case).

pweil- commented:
https://github.com/kubernetes/kubernetes/issues/35983

@k8s-bot node e2e test this

smarterclayton commented:
@k8s-bot cvm gce e2e test this

Not sure what the issue is but no output from GCE on failure

smarterclayton commented:
@k8s-bot gci gke e2e test this

pweil- commented:
rebased

pweil- commented:
rebased

pweil- commented:
@k8s-bot unit test this

https://github.com/kubernetes/kubernetes/issues/36351

pweil- commented:
@k8s-bot cvm gce e2e test this

pweil- commented:
rebased tests are green again, care to re-apply LGTM?

@smarterclayton @derekwaynecarr

pweil- commented:
@vishh ^

pweil- commented:
@vishh bump

pweil- commented:
@saad-ali FYI this has had LGTM and has been in rebase cycles since (if you need to do any release czar maintenance)

saad-ali commented:
> @saad-ali FYI this has had LGTM and has been in rebase cycles since (if you need to do any release czar maintenance)

Ack. Thanks for the heads up. Are there any concerns about this PR in terms of risk to the 1.5 release? I see that the initial plan was to get it merged early in the 1.5 dev cycle to give it time to bake.

vishh commented:
Saad, this PR is expected to be turned off by default. So from a risk
perspective, it should be low.

On Tue, Nov 8, 2016 at 12:27 PM, Saad Ali <notifications@github.com> wrote:

> @saad-ali <https://github.com/saad-ali> FYI this has had LGTM and has
> been in rebase cycles since (if you need to do any release czar maintenance)
>
> Ack. Thanks for the heads up. Are there any concerns about this PR in
> terms of risk to the 1.5 release? I see that the initial plan was to get it
> merged early in the 1.5 dev cycle to give it time to bake.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31169#issuecomment-259249885>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKEq31lgRSY-p5m22zJLbQlGEvFMIks5q8NtOgaJpZM4JqWln>
> .
>

pweil- commented:
> Saad, this PR is expected to be turned off by default. So from a risk
> perspective, it should be low.

Agree, the feature is guarded by a feature gate and is turned off by default

saad-ali commented:
>> Saad, this PR is expected to be turned off by default. So from a risk
>> perspective, it should be low.
>
> Agree, the feature is guarded by a feature gate and is turned off by default

Great, thanks.

deads2k commented:
> Saad, this PR is expected to be turned off by default. So from a risk
> perspective, it should be low.
> Agree, the feature is guarded by a feature gate and is turned off by default
> Great, thanks.

Looks like this has approval.  Retagging.

dims commented:
@deads2k @saad-ali - looks like one of you have to reapply lgtm tag as the bot removed it

pweil- commented:
> @deads2k @saad-ali - looks like one of you have to reapply lgtm tag as the bot removed it

yes, please.  PR needed rebase.  Thanks!
